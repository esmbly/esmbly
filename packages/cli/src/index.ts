import yargs, { Arguments } from 'yargs';
import * as init from './commands/init';
import * as run from './commands/run';

export function runCLI(): Arguments {
  return yargs
    .scriptName('esmbly')
    .command(init)
    .command(run)
    .help().argv;
}
