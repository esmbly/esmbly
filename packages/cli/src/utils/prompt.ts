import inquirer from 'inquirer';
import {
  TransformerInfo,
  getAvailableOutputFormats,
  getAvailableTransformers,
} from '@esmbly/utils';
import stringify from 'stringify-object';
import chalk from 'chalk';

interface PromptedConfig {
  module: string;
  install: boolean;
  transformers: TransformerInfo[];
  pkgManager: string;
}

function createRequireString(transformers: TransformerInfo[]): string {
  return transformers
    .map((t: TransformerInfo) => `const ${t.const} = require('${t.pkg}')`)
    .join('\n');
}

function createConfigString(
  input: string,
  output: string[],
  outDir: string,
): string {
  return stringify({
    input: [input],
    transformers: '<TRANSFORMERS>',
    // eslint-disable-next-line
    output: output.map((format: string) => ({ format, outDir })),
  });
}

function createTransformerCalls(transformers: TransformerInfo[]): string {
  /**
   * Produces:
   * [
   *   Flow.createTransformer(),
   *   Wasm.createTransformer()
   * ]
   */
  return `${transformers.reduce(
    (calls, t) => `${calls}\n    ${t.const}.createTransformer(),`,
    '[',
  )}\n  ]`;
}

function createModuleString(
  input: string,
  transformers: TransformerInfo[],
  output: string[],
  outDir: string,
): string {
  const requireString = createRequireString(transformers);
  const configString = createConfigString(input, output, outDir);
  const transformerCalls = createTransformerCalls(transformers);

  return `${requireString}\n\nmodule.exports = ${configString.replace(
    `'<TRANSFORMERS>'`,
    transformerCalls,
  )}\n`;
}

function isValidCombination(transformers: TransformerInfo[]): boolean {
  let isValid = true;

  for (let i = 0; i < transformers.length - 1; i += 1) {
    const current = transformers[i];
    const next = transformers[i + 1];

    if (!current.to.includes(next.from)) {
      isValid = false;
      break;
    }
  }

  return isValid;
}

export async function promptForConfig(): Promise<PromptedConfig> {
  const availableTransformers = getAvailableTransformers();

  const { input } = await inquirer.prompt([
    {
      message: 'Which files do you want to transform?',
      name: 'input',
      suffix: chalk.dim(' e.g ./**/*.js'),
      type: 'input',
    },
  ]);

  const { selectedTransformers } = await inquirer.prompt([
    {
      choices: availableTransformers.map(t => ({
        name: `${t.name}\t${chalk.dim(`${t.from} -> ${t.to.join(' | ')}`)}`,
        short: t.name,
        value: t.name,
      })),
      default: [],
      message: 'Which transformers do you want to use?',
      name: 'selectedTransformers',
      type: 'checkbox',
    },
  ]);

  const transformers = availableTransformers.filter(t =>
    selectedTransformers.includes(t.name),
  );

  if (!isValidCombination(transformers)) {
    const combination = transformers.reduce(
      (str: string, transformer: TransformerInfo, i: number) => {
        if (i === 0) {
          return `${transformer.name}`;
        }

        return `${str} -> ${transformer.name}`;
      },
      '',
    );
    throw new Error(
      `Transformations in the order ${combination} is not possible`,
    );
  }

  const { output } = await inquirer.prompt([
    {
      choices: getAvailableOutputFormats(transformers),
      message: 'Which formats to you want to output?',
      name: 'output',
      type: 'checkbox',
      when: transformers.length > 0,
    },
  ]);

  const { outDir } = await inquirer.prompt([
    {
      message: 'To which directory do you want to output files?',
      name: 'outDir',
      suffix: chalk.dim(' e.g ./dist'),
      type: 'input',
    },
  ]);

  const { install } = await inquirer.prompt([
    {
      default: true,
      message: 'Do you want to install the selected transformers and the CLI?',
      name: 'install',
      type: 'confirm',
      when: transformers.length > 0,
    },
  ]);

  const { pkgManager } = await inquirer.prompt([
    {
      choices: ['npm', 'yarn'],
      message: 'Which package manager do you want to use for installation?',
      name: 'pkgManager',
      type: 'list',
      when: install === true,
    },
  ]);

  return {
    install,
    module: createModuleString(input, transformers, output, outDir),
    pkgManager,
    transformers,
  };
}
