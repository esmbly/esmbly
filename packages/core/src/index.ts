import { File, RunConfig } from '@esmbly/types';
import { parse } from './parse';
import { validateConfig, validateInputFormat } from './validate';

export async function run(config: RunConfig): Promise<File[]> {
  validateConfig(config);

  let files: File[] = [];
  const trees = parse(config.input, config.transformers[0]);

  for (const transformer of config.transformers) {
    if (transformer.before) {
      await transformer.before();
    }

    validateInputFormat(trees, transformer);

    if (transformer.transform) {
      await transformer.transform(trees);
    }

    if (transformer.createFiles) {
      files = [
        ...files,
        ...(await transformer.createFiles(trees, config.output)),
      ];
    }

    if (transformer.after) {
      await transformer.after();
    }
  }

  return files;
}
