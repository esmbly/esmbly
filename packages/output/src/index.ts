type Callback = (error: Error | null | undefined) => void;
type Encoding = string | undefined;
type WriteFn = (
  chunk: string,
  encoding?: Encoding,
  callback?: Callback,
) => boolean;

class Output {
  private stdout: NodeJS.WriteStream | null;

  private stderr: NodeJS.WriteStream | null;

  public constructor() {
    this.stdout = process ? process.stdout : null;
    this.stderr = process ? process.stderr : null;
  }

  public get isTTY(): boolean {
    if (this.stdout) {
      return this.stdout.isTTY === true;
    }
    return false;
  }

  public err(chunk: string, encoding?: Encoding, callback?: Callback): boolean {
    if (this.stderr) {
      return this.stderr.write(chunk, encoding, callback);
    }
    return false;
  }

  public out(chunk: string, encoding?: Encoding, callback?: Callback): boolean {
    if (this.stdout) {
      return this.stdout.write(chunk, encoding, callback);
    }
    return false;
  }

  public clearLine(...args: WriteFn[]): void {
    if (this.isTTY) {
      args.forEach(fn => fn.call(this, '\x1b[999D\x1b[K'));
    }
  }

  public setErrorStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stderr = stream as NodeJS.WriteStream;
  }

  public setOutStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stdout = stream as NodeJS.WriteStream;
  }
}

export default new Output();
