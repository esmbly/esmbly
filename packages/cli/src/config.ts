import path from 'path';
import stringify from 'stringify-object';
import { Config } from '@esmbly/types';
import {
  writeFile,
  readFile,
  exists,
  getRoot,
  getRelativePathTo,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';

export const DEFAULT_FILE = '.esmblyrc.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(__dirname, `../defaults/${DEFAULT_FILE}`);
  const template = await readFile(templatePath);
  return template.toString();
}

export async function getDefaultConfigPath(): Promise<string> {
  const root = await getRoot();
  return path.join(root, DEFAULT_FILE);
}

export async function readConfig(customPath?: string): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = require(configPath) // eslint-disable-line
  const toArray = (maybeArrayConfig: Config | Config[]): Config[] => {
    if (Array.isArray(maybeArrayConfig)) {
      return maybeArrayConfig;
    }
    return [maybeArrayConfig];
  };
  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArray(configFnResult);
  }
  return toArray(config);
}

export async function createConfig(
  useDefault: boolean,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = await getRoot();
  const defaultConfigPath = await getDefaultConfigPath();
  if (await exists(defaultConfigPath)) {
    const relativeConfigPath = await getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (useDefault) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template);
  } else {
    const config = await promptForConfig();
    const content = `module.exports = ${stringify(config)}\n`;
    await writeFile(defaultConfigPath, content);
  }
  return { root, fileName: DEFAULT_FILE };
}
