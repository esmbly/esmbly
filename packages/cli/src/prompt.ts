import inquirer from 'inquirer';
import { Config } from '@esmbly/types';
import { getTransformers, getOutputFormats } from '@esmbly/utils';

export async function promptForConfig(): Promise<Config> {
  const { input } = await inquirer.prompt([
    {
      name: 'input',
      message: 'Which files do you want to transform?',
      type: 'input',
    },
  ]);

  const { transformers } = await inquirer.prompt([
    {
      name: 'transformers',
      message: 'Which transformer do you want to use?',
      type: 'checkbox',
      choices: await getTransformers(),
    },
  ]);

  // TODO: Ask to install transformers here?

  const { output } = await inquirer.prompt([
    {
      name: 'output',
      message: 'Which output formats to you want to use?',
      type: 'checkbox',
      choices: await getOutputFormats(transformers),
    },
  ]);

  return { input: [input], transformers, output };
}
