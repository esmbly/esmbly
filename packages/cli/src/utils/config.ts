import path from 'path';
import { Config, InitOptions } from '@esmbly/types';
import {
  exists,
  getRelativePathTo,
  getRoot,
  readFile,
  writeFile,
} from '@esmbly/utils';
import { promptForConfig } from './prompt';
import { installCLI, installTransformers } from './install';

export const DEFAULT_FILE = 'esmbly.config.js';

export async function getTemplateConfig(): Promise<string> {
  const templatePath = path.resolve(
    __dirname,
    `../../defaults/${DEFAULT_FILE}`,
  );
  const template = await readFile(templatePath);
  return template.toString();
}

export function getDefaultConfigPath(): string {
  const root = getRoot();
  return path.join(root, DEFAULT_FILE);
}

type ConfigFnA = () => Config;
type ConfigFnB = () => Config[];
type ConfigFn = ConfigFnA | ConfigFnB;

function toArrayConfig(maybeArrayConfig: Config | Config[]): Config[] {
  if (Array.isArray(maybeArrayConfig)) {
    return maybeArrayConfig;
  }

  return [maybeArrayConfig];
}

export async function readConfig(
  customPath?: string,
  requirer: (requirePath: string) => unknown = require,
): Promise<Config[]> {
  const configPath = customPath || (await getDefaultConfigPath());
  const config = requirer(configPath) as Config | Config[] | ConfigFn;

  if (typeof config === 'function') {
    const configFnResult = await config();
    return toArrayConfig(configFnResult);
  }

  return toArrayConfig(config);
}

export async function createConfig(
  options: InitOptions,
): Promise<{
  root: string;
  fileName: string;
}> {
  const root = getRoot();
  const defaultConfigPath = getDefaultConfigPath();

  if ((await exists(defaultConfigPath)) && !options.force) {
    const relativeConfigPath = getRelativePathTo(defaultConfigPath);
    throw new Error(
      `Config ${relativeConfigPath} already exists. Remove that file first.`,
    );
  } else if (options.default) {
    const template = await getTemplateConfig();
    await writeFile(defaultConfigPath, template, { overwrite: options.force });
  } else {
    const config = await promptForConfig();

    if (config.install) {
      await installTransformers(config.transformers, config.pkgManager);
      await installCLI(config.pkgManager);
    }

    await writeFile(defaultConfigPath, config.module, {
      overwrite: options.force,
    });
  }

  return { fileName: DEFAULT_FILE, root };
}
