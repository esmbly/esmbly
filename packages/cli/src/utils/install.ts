import { exec } from 'child_process';
import { promisify } from 'util';
import { TransformerInfo } from '@esmbly/utils';
import ora from 'ora';

export function getPkgManagerCommand(pkgManager: string): string {
  switch (pkgManager) {
    case 'npm':
      return 'npm install --save-dev';
    case 'yarn':
      return 'yarn add --dev';
    default:
      throw new Error(`Package manager ${pkgManager} is not supported`);
  }
}

export async function installTransformers(
  transformers: TransformerInfo[],
  pkgManager: string,
): Promise<void> {
  const command = getPkgManagerCommand(pkgManager);

  for (const transformer of transformers) {
    const spinner = ora(`Installing: ${transformer.pkg}`).start();
    await promisify(exec)(`${command} ${transformer.pkg}`);
    spinner.succeed(`Installed: ${transformer.pkg}`);
  }
}

export async function installCLI(pkgManager: string): Promise<void> {
  const command = getPkgManagerCommand(pkgManager);

  const spinner = ora('Installing: @esmbly/cli').start();
  await promisify(exec)(`${command} @esmbly/cli`);
  spinner.succeed('Installed: @esmbly/core');
}
