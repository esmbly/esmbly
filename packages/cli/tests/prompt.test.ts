import { promptForConfig } from '../src/prompt';
import * as utils from '@esmbly/utils';
import inquirer from 'inquirer';

jest.mock('inquirer');
jest.mock('@esmbly/utils');

describe('promptForConfig', () => {
  it('asks the correct questions', async () => {
    inquirer.prompt = jest.fn();
    (inquirer.prompt as jest.Mock).mockReturnValue({});
    (utils as any).getTransformers = jest.fn();
    (utils.getTransformers as jest.Mock).mockResolvedValue(['flow', 'wasm']);
    (utils as any).getOutputForTransformers = jest.fn();
    (utils.getOutputFormats as jest.Mock).mockResolvedValue([
      'TypeScript',
      'WebAssembly',
    ]);
    await promptForConfig();
    expect((inquirer.prompt as jest.Mock).mock.calls).toMatchSnapshot();
    expect(utils.getTransformers as jest.Mock).toHaveBeenCalledTimes(1);
    expect(utils.getOutputFormats as jest.Mock).toHaveBeenCalledTimes(1);
  });
});
