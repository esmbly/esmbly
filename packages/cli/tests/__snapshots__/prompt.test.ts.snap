// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`promptForConfig asks the correct questions 1`] = `
Array [
  Array [
    Array [
      Object {
        "message": "Which files do you want to transform?",
        "name": "input",
        "type": "input",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "flow",
          "wasm",
        ],
        "message": "Which transformer do you want to use?",
        "name": "transformers",
        "type": "checkbox",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "TypeScript",
          "WebAssembly",
        ],
        "message": "Which output formats to you want to use?",
        "name": "output",
        "type": "checkbox",
      },
    ],
  ],
]
`;
